app/javascript/packages/i18n/index.ts

```diff
@@ -0,0 +1,83 @@
+interface PluralizedEntry {
+  one: string;
+  other: string;
+}
+
+type Entry = string | PluralizedEntry;
+type Entries = Record<string, Entry>;
+type Variables = Record<string, any>;
+
+interface I18nOptions {
+  strings?: Entries;
+}
+
+const hasOwn = (object: object, key: string): boolean =>
+  Object.prototype.hasOwnProperty.call(object, key);
+
+/**
+ * Returns the pluralization object key corresponding to the given number.
+ *
+ * @param count Count.
+ *
+ * @return Pluralization key.
+ */
+const getPluralizationKey = (count?: number): keyof PluralizedEntry =>
+  count === 1 ? 'one' : 'other';
+
+/**
+ * Returns an entry from locale data.
+ *
+ * @param entry Locale data.
+ * @param count Pluralization count, if applicable.
+ *
+ * @return Entry string or object.
+ */
+const getEntry = (strings: Entries, key: string): Entry =>
+  hasOwn(strings, key) ? strings[key] : key;
+
+/**
+ * Returns the resulting string from the given entry, incorporating pluralization if necessary.
+ *
+ * @param entry Entry string or object.
+ * @param count Pluralization count, if applicable.
+ *
+ * @return Entry string.
+ */
+const getString = (entry: Entry, count?: number): string =>
+  typeof entry === 'object' ? entry[getPluralizationKey(count)] : entry;
```

Comment:
should we also make sure `count` in not undefined? Otherwise the combo of a too-short key and a missing count 
could result in the wrong value?
```suggestion
  typeof entry === 'object' && count !== undefined ? entry[getPluralizationKey(count)] : entry;
```

Reply:
If count is an optional param that is only supplied if we need to call getPluralizationKey, why not simplify to 
```count != undefined ? entry[getPluralizationKey(count)] : entry;```

Reply:
To this and @SammySteiner 's point below, I suppose there's a strong connection between `count` and `PluralizedEntry`. It'd be nice if we could define it in a way that TypeScript would enforce it on the consumer end, though I dunno if that'd be possible.

Reply:
I was imagining like a case where we have strings like this:

```
some: {
  value: {
    a: "the string"
  }
}
```
and if we call `t('some.value')` in Ruby i18n we get the whole hash: `{ a: "the string" }` so my worry here was that if we did that, the JS code here would be like "ah yes, an object let me look up the count"

A TypeError seems appropriate to me, it also depends how much we want to match Ruby i18n. I think that returning a whole object is rarely desirable behavior so we could just TypeError on that until we find ourselves needing that.

Reply:
> and if we call `t('some.value')` in Ruby i18n we get the whole hash: `{ a: "the string" }` so my worry here was that if we did that, the JS code here would be like "ah yes, an object let me look up the count"
> 
> A TypeError seems appropriate to me, it also depends how much we want to match Ruby i18n. I think that returning a whole object is rarely desirable behavior so we could just TypeError on that until we find ourselves needing that.

That's a good point. And while I agree it's not really a common use-case, if it's easy enough to match the behavior, I'll try to do that.

Reply:
I went down the route of trying to match the Ruby behavior, but ultimately I think it would be more hassle than it's worth for how it would impact usage, since it would change the return value of `t` from `string` to `Entry` (i.e. a string or an object), which would make usage more cumbersome when 99%+ of use-cases would be to use the `t` result as a string.

Instead, I went with the idea to throw an error when missing a `count` for an object entry in f3fbfc182.